src/leveled_sst.erl

```diff
@@ -1725,20 +1764,107 @@ hmac(Int) when is_integer(Int) ->
 %%
 %% This implementation of the SlotIndex uses leveled_tree
 
-from_list(SlotList) ->
-    leveled_tree:from_orderedlist(SlotList, ?TREE_TYPE, ?TREE_SIZE).
+from_list(SlotList, FirstKey, LastKey) ->
+    FilterFun = get_filterfun(FirstKey, LastKey),
+    FilteredList =
+        lists:map(fun({K, S}) -> {FilterFun(K), S} end, SlotList),
+    {leveled_tree:from_orderedlist(FilteredList, ?TREE_TYPE, ?TREE_SIZE),
+        FilterFun}.
+
+-spec get_filterfun(
+    leveled_codec:ledger_key(), leveled_codec:ledger_key()) ->
+        fun((leveled_codec:ledger_key())
+            -> leveled_codec:ledger_key()|leveled_codec:slimmed_key()).
+get_filterfun(
+    {Tag, Bucket, {Field, Term}, FK}, {Tag, Bucket, {Field, Term}, LK})
+            when is_binary(Term), is_binary(FK), is_binary(LK) ->
+    subkey_filter(Term);
+get_filterfun(
+        {Tag, Bucket, {Field, FT}, FK}, {Tag, Bucket, {Field, LT}, LK})
+            when is_binary(FT), is_binary(FK), is_binary(LT), is_binary(LK) ->
+    case {binary:longest_common_prefix([FT, LT]), byte_size(FT)} of
+        {N, M} when N > 0, M > N ->
+            <<Prefix:N/binary, _Rest/binary>> = FT,
+            term_prefix_filter(N, Prefix);
+        _ ->
+            fun term_filter/1
+    end;
+get_filterfun(
+        {Tag, Bucket, FK, null}, {Tag, Bucket, LK, null})
+            when is_binary(FK), is_binary(LK), FK < LK ->
+    case {binary:longest_common_prefix([FK, LK]), byte_size(FK)} of
+        {N, M} when N > 0, M > N ->
+            <<Prefix:N/binary, _Rest/binary>> = FK,
+            key_prefix_filter(N, Prefix);
+        _ ->
+            fun key_filter/1
```

Comment:
I can't quite remember the rules on ordering, but is it possible to have FK="0000" and LK="0000999999"? or does "0000" come after? if it's the first, then I don't think this would result in an optimised FilterFun. In the 2nd case, this may need a test case to ensure expected behaviour (there may be one already I've not spotted)

Reply:
So <<"0000">> is smaller than <<"0000">> with anything appended so <<"0000999999">> is > than <<"0000">>.

I think the fact that not every case is optimised is fine, it isn't going to have a material impact if the occasional SST file is using more memory than is strictly required.  So I would rather have simpler code than have full optimisation.

The unit test `key_matchesprefix_test/0` In both the index case and object case creates a First Key that is equal in size to the longest_common_prefix.  But I perhaps added this test as I assumed it would be optimised, when no in fact it just becomes a `key_filter/1` test.

Need to clear my head of other things, and just think this through carefully again.

Reply:
I'm sure M == N is ok.  The guard M > N was introduced before the Prefix check was added to the filter functions (we were just truncating for size).  The weirdness that might have been caused in these cases is now handled by setting to null whenever the first N don't match the Prefix.

I've updated in https://github.com/martinsumner/leveled/pull/383/commits/0fe735fa823daf9f92e2d468783fbec64483bd49 and all tests pass.